### ISSUE TITLE
[RFC]: Isolate OpenAI Server Into Separate Process

### ISSUE THREAD

**Issue by robertgshaw2-neuralmagic** [vllm-project/vllm#6797]
### Motivation.

Currently, the OpenAI API server and AsyncLLMEngine share the same asyncio event loop. This means that the API server and the CPU components of the AsyncLLMEngine contend for the same resources. Below, we have a chart of Llama-3-8B running ShareGPT at QPS=10 on an H100.

We can see time is split into three buckets:
- Light blue 	→ this is overlapped GPU execution time (call of execute_model)
- Orange 	→ this is CPU execution time in the LLM engine
- All else		→ this is API server time

<img width="851" alt="image" src="https://github.com/user-attachments/assets/96cbc8c5-7fb8-4c48-b40a-9aa00108b1df">

- `execute_model` does not block the asyncio event loop as it is run in run_in_execuctor https://github.com/vllm-project/vllm/blob/d92b3c5cdef59533347ac714a70274f186943019/vllm/executor/gpu_executor.py#L143
- So, we believe that some GPU time is currently already overlapped with the API server

### Proposed Change.

Enable better overlapping of the API server and `AsyncLLMEngine` via multiprocessing by spitting into two processes that communicate over gRPC with `protobufs`. This is roughly the architecture used by TGI (though TGI has more items in the server than we are proposing here.

### Initial Goal

- Write protobuf API to roughly match AsyncLLMEngine `generate()` method, separate API server into separate process and measure the speedup
    - Will exclude LogitsProcessors - it’s agreed that these don’t belong in SamplingParams anyhow. Instead we’ll include the corresponding external API parameters e.g. for guided decoding
    - Can be a gRPC streaming response, which maps nicely to the stream returned by this method. Cancellations can be mapped to request aborts

### Follow Ups
Move items that currently run inside `AsyncLLMEngine` into the API Server for better overlap with the GPU

- Make necessary changes within the engine to allow full token-id/token-id out so that it can be called such that no tokenization/detokenization is done - you can already pass token ids but there are some places where tokenization/detokenization is done unconditionally such as detokenizing the prompt to return
- Decouple the LogitsProcessor construction from the OpenAI API code - into layer/utility for creating them from the top-level parameters. This can then be called from the gRPC API layer to construct the full SamplingParams
- Decouple incremental detokenization logic so that it can also be used in the API server process
- The API server process can also intercept/process any stop strings, and just abort the request as needed

### CC List.

@simon-mo @njhill @dsikka @joerunde

**Comment by mgoin** [comment]
Outside of gRPC, [ZMQ](https://zeromq.org/) could be a suitable lightweight alternative that is already used inside vLLM: https://github.com/vllm-project/vllm/blob/6a1e25b1514a25d3da96d0d78c4568f6e581e242/requirements-common.txt#L24 

Initial PR usage for reference: https://github.com/vllm-project/vllm/pull/6183

**Comment by zifeitong** [comment]
> Outside of gRPC, [ZMQ](https://zeromq.org/) could be a suitable lightweight alternative that is already used inside vLLM:
> 
> https://github.com/vllm-project/vllm/blob/6a1e25b1514a25d3da96d0d78c4568f6e581e242/requirements-common.txt#L24
> 
> Initial PR usage for reference: #6183

+1, it's worth doing some benchmark. python_grpc [isn't that performant](https://github.com/LesnyRumcajs/grpc_bench/discussions/441).

Triton Inference Server uses [shared memory](https://github.com/triton-inference-server/python_backend/tree/main?tab=readme-ov-file#managing-shared-memory) for commutation.

**Comment by robertgshaw2-neuralmagic** [comment]
Sounds good. We are working on a prototype and will measure

shouldn’t be too hard to swap out protocols in the prototype since it’s very simple

**Comment by K-Mistele** [comment]
Would it be accurate to say that the global interpreter lock is the root cause of the issue here?

**Comment by simon-mo** [comment]
> Would it be accurate to say that the global interpreter lock is the root cause of the issue here?
Pretty much. If we can do freely threaded Python the problem will go away, but the it is still work in progress.

**Comment by robertgshaw2-neuralmagic** [comment]
> > Would it be accurate to say that the global interpreter lock is the root cause of the issue here?
> 
> Pretty much. If we can do freely threaded Python the problem will go away, but the it is still work in progress.

Its not quite as bad as the chart shows, since GPU execution releases the GIL + some of the front end is overlapped. But its hard to tell exactly how much thus the experiments we are trying

gRPC is one approach, but there are other ways to add multiprocessing if this does not work

**Comment by pzhao1799** [comment]
> +1, it's worth doing some benchmark. python_grpc [isn't that performant](https://github.com/LesnyRumcajs/grpc_bench/discussions/441).

In fact, in grpc's own [documentation](https://grpc.io/docs/guides/performance/#python), they claim that streaming RPCs in python is even less performant than unary RPCs. In my experience, this is roughly true so I generally avoid grpc in python if possible. I think another protocol or leveraging shm are all good things to try, but getting some benchmarked values of grpc (if that's already in the works) will give us somewhere to start.

**Comment by robertgshaw2-neuralmagic** [comment]
Yeah, it’s a simple prototype to see what the best approach is. Early experiments with grpc aren’t looking great so I’m going to try out zmq tomorrow

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks to everyone for the feedback! We made three prototypes, separating into separate processes:
- `grpc` based -> https://github.com/vllm-project/vllm/pull/6839
- `zeromq` based with python pickle -> https://github.com/vllm-project/vllm/pull/6839
- `zeromq` based with protobufs ->

TLDR is that the `zeromq` based protocol is delivering ~20% speedup at 10QPS for Llama-3-8B-Instruct on 1xH100 (550 input tokens | 150 output tokens)


| Metric | `main` | `grpc` | `zmq + pickle` | `zmq + protobuf` |
| - | - | - | - | - |
| TPOT (mean) | 70.08 | 77.93 | 57.35 | 54.54 |
| TTFT (mean) | 273.77 | 99.31 | 298.09 | 269.01 |

**Comment by rockwotj** [comment]
Another framework to consider is ucall: https://github.com/unum-cloud/ucall

**Comment by vrdn-23** [comment]
Apologies if this is not relevant to this issue. Is there a particular reason why vLLM has not considered moving the app server to [gunciorn](https://gunicorn.org) instead of uvicorn? 

Uvicorn doesn't do process management and is not recommended to be used for production-grade systems ([as mentioned in the uvicorn documentation itself](https://www.uvicorn.org/#running-with-gunicorn)).

 I can see that a similar issue was raised at one point (#2573) but it did not seem to get much traction. I was just wondering if this was something that the maintainer team has considered as an option or is this something that can be prioritized hopefully?

**Comment by robertgshaw2-neuralmagic** [comment]
> Apologies if this is not relevant to this issue. Is there a particular reason why vLLM has not considered moving the app server to [gunciorn](https://gunicorn.org) instead of uvicorn?
> 
> Uvicorn doesn't do process management and is not recommended to be used for production-grade systems ([as mentioned in the uvicorn documentation itself](https://www.uvicorn.org/#running-with-gunicorn)).
> 
> I can see that a similar issue was raised at one point (#2573) but it did not seem to get much traction. I was just wondering if this was something that the maintainer team has considered as an option or is this something that can be prioritized hopefully?

We will certainly evaluate `guincorn` in the follow ups to this effort

**Comment by simon-mo** [comment]
Previously we only need to launch a single process. When we move towards multiple server processes, gunicorn is definitely the go to option.

**Comment by richarddli** [comment]
I think that whether or not to include a WSGI server in the stack like Gunicorn in addition to uvicorn depends on the deployment model. I deploy vLLM into a Kubernetes cluster, and am perfectly happy with uvicorn, and I use Kubernetes for all process management. I could see how Gunicorn might be helpful if a common deployment model is on VMs to enable more efficient use of the cores (?). But my bias would be leave the scale up deployment model up to the user who's deploying it.

**Comment by vrdn-23** [comment]
> But my bias would be leave the scale up deployment model up to the user who's deploying it.

I think that makes sense to me. I think an acceptable compromise would then have something like a factory function of some sorts that returns the fully initialized FastAPI application so that we can invoke it to wrap it around uvicorn/gunicorn as we please. 

My understand in this is limited, but I think a similar refactor to what we have in #7095 will probably work? I'm not sure if gunicorn allows to pass arbitrary arguments into a function that defines the api, but if we make args and the engine arguments global, it might work? Any particular thoughts on this @DarkLight1337 ?

**Comment by DarkLight1337** [comment]
> > But my bias would be leave the scale up deployment model up to the user who's deploying it.
> 
> I think that makes sense to me. I think an acceptable compromise would then have something like a factory function of some sorts that returns the fully initialized FastAPI application so that we can invoke it to wrap it around uvicorn/gunicorn as we please.
> 
> My understand in this is limited, but I think a similar refactor to what we have in #7095 will probably work? I'm not sure if gunicorn allows to pass arbitrary arguments into a function that defines the api, but if we make args and the engine arguments global, it might work? Any particular thoughts on this @DarkLight1337 ?

If you want to use gunicorn, you can create a new file and use `init_app` from the original entrypoints file to expose the ASGI app object, then run gunicorn on that file.

**Comment by jon-chuang** [comment]
> When we move towards multiple server processes, gunicorn is definitely the go to option.

Is the idea to have multiple API server processes as the frontend that each talk to the main AsyncLLMEngine process?

**Comment by robertgshaw2-neuralmagic** [comment]
> > When we move towards multiple server processes, gunicorn is definitely the go to option.
> 
> Is the idea to have multiple API server processes as the frontend that each talk to the main AsyncLLMEngine process?
> 
> And to offload more computational work (e.g. detokenize) into such processes so as to reduce the CPU time on the main process?

Correct

**Comment by jon-chuang** [comment]
@robertgshaw2-neuralmagic do you have an estimate of how much work can be split out from the AsyncLLMEngine into the frontend processes based on your profiling?

**Comment by robertgshaw2-neuralmagic** [comment]
Just detokenization

**Comment by njhill** [comment]
And stopping evaluation (stop string checking etc.)

**Comment by jon-chuang** [comment]
> And stopping evaluation (stop string checking etc.)

Could you elaborate? Is this EOS? Isn't this done in scheduler?

> Just detokenization

What's the est. fraction of the AsyncLLMEngine CPU time?

Why does this require > 1 process unless dealing with large batch size?

**Comment by daquexian** [comment]
Inspired by this RFC, why not move heavy CPU tasks into a separate process as much as possible for maximum CPU-GPU overlap? I noticed there are some efforts like https://github.com/vllm-project/vllm/issues/6854 and https://github.com/vllm-project/vllm/issues/6913 to help the overlap but IMO they are tricky and possibly incompatible with other features like chunked prefill.

**Comment by njhill** [comment]
> > And stopping evaluation (stop string checking etc.)
> 
> Could you elaborate? Is this EOS? Isn't this done in scheduler?
>

@jon-chuang yes of course EOS and max length checking is trivial and thinking to remain where it is now. Was referring primarily to stop string evaluation, and possibly stop token ids. Although there is some tradeoff once for non-streaming requests once https://github.com/vllm-project/vllm/pull/7381 is included since we'll have to stream the responses to the other process that we wouldn't have had to otherwise.
 
> > Just detokenization
> 
> What's the est. fraction of the AsyncLLMEngine CPU time?
> 
> Why does this require > 1 api server process unless dealing with large batch size?

None of this requires > 1 api server process. It should be easy to have more than one but I don't think we'll need to.

> Inspired by this RFC, can we move heavy CPU tasks into separate processes as much as possible
@daquexian yes, this is started in https://github.com/vllm-project/vllm/pull/7402.

**Comment by JoanFM** [comment]
Hello,

Out of curiosity, have you considered simply using ThreadPools to unblock the event loop if the CPU time consuming I assume is mostly taken by libraries (such as tokenizers) that would release the GIL?

The communication may be easier and faster as trade for some GIL locking?

Thanks

**Comment by robertgshaw2-neuralmagic** [comment]
> Out of curiosity, have you considered simply using ThreadPools to unblock the event loop if the CPU time consuming I assume is mostly taken by libraries (such as tokenizers) that would release the GIL?

Responded in #8033

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks to everyone for the help! `zeromq` is very useful for mulitprocess python!
